keps/sig-node/4033-group-driver-detection-over-cri/README.md

```diff
@@ -0,0 +1,890 @@
+conversion tests if API types are being modified.
+
+Additionally, for features that are introducing a new API field, unit tests that
+are exercising the `switch` of feature gate itself (what happens if I disable a
+feature gate after having objects written with the new field) are also critical.
+You can take a look at one potential example of such test in:
+https://github.com/kubernetes/kubernetes/pull/97058/files#diff-7826f7adbc1996a05ab52e3f5f02429e94b68ce6bce0dc534d1be636154fded3R246-R282
+-->
+
+N/A.
+
+### Rollout, Upgrade and Rollback Planning
+
+<!--
+This section must be completed when targeting beta to a release.
+-->
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+<!--
+Try to be as paranoid as possible - e.g., what if some components will restart
+mid-rollout?
+
+Be sure to consider highly-available clusters, where, for example,
+feature flags will be enabled on some API servers and not others during the
+rollout. Similarly, consider large clusters and how enablement/disablement
+will rollout across nodes.
+-->
+
+A rollout/rollback could fail only in the way it currently does: cgroup driver
+skew between CRI server (containe runtime) and Kubelet resulting in nodes going
+NotReady. This is only possible when the CRI server and Kubelet are not both
+upgraded to support the feature and are both not configured to agree on the
+CgroupDriver as they must be today.
+
+###### What specific metrics should inform a rollback?
+
+<!--
+What signals should users be paying attention to when the feature is young
+that might indicate a serious problem?
+-->
+
+N/A.
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+<!--
+Describe manual testing that was done and the outcomes.
+Longer term, we may want to require automated upgrade/rollback tests, but we
+are missing a bunch of machinery and tooling and can't do that now.
+-->
+
+N/A.
+
+###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+<!--
+Even if applying deprecation policies, they may still surprise some users.
+-->
+
+Yes, the CgroupDriver field of the Kubelet configuration (and the corresponding
+`--cgroup-driver` flag) will be marked as deprecated, but won’t be dropped
+until wide adoption of the field on CRI has been done.
+
+### Monitoring Requirements
+
+<!--
+This section must be completed when targeting beta to a release.
+
+For GA, this section is required: approvers should be able to confirm the
+previous answers based on experience in the field.
+-->
+
+###### How can an operator determine if the feature is in use by workloads?
+
+<!--
+Ideally, this should be a metric. Operations against the Kubernetes API (e.g.,
+checking if there are objects with field X set) may be a last resort. Avoid
+logs or events for this purpose.
+-->
+
+Kubelet and CRI version.
```

Comment:
If I know these values, and have no information about the container runtime, can I work out whether or not that the kubelet is automatically learning which cgroup driver to use?

Reply:
Woops, this should actually state "container runtime" instead of "CRI". PR updated accordingly.

The end user doesn't need to know about this. The node admin ahould know what version of container runtime is in use. @haircommander

Reply:
For Kubernetes, the node admin is usually _our_ end user. We use the term “cluster operator” in docs (which does sometimes confuse, but that's what we have in the style guide).

Reply:
Ach, yes of course, thanks for correcting me. But anyway, the answer would be "Kubelet and container runtime version"

Reply:
This now states:

_Kubelet and container runtime version. The [`crictl`](https://kubernetes.io/docs/tasks/debug/debug-cluster/crictl/) tool can be used to determine if the container runtime supports the feature (`crictl info`)._

@sftim do you consider this as resolved?

Reply:
Resolving this comment as a duplicate of https://github.com/kubernetes/enhancements/pull/4034#discussion_r1217944720